### DISCUSSION TITLE
RPC Return Codes when used as a library

### DISCUSSION THREAD

**Discussion by rsdmike** [open-amt-cloud-toolkit/open-amt-cloud-toolkit#146] [General]
For those interested in using RPC-GO in a library. Here are the return codes we are thinking. If there are any missing or have any thoughts, please let us know.

TL:DR
0: Success
1-99: General Errors using RPC
100-199: AMT specific error messages
200+: future technology specific error message.  Each technology will occupy a set of 100 error slots (200-299, 300-399, etc)

### (1-19) Basic errors outside of Open AMT Cloud Toolkit
1: Incorrect permissions (not admin or sudo)
2: HECI driver not detected
3: AMT not detected
4: AMT not ready

### (20-69) Input errors to RPC
20: Missing or incorrect URL
21: Missing or incorrect profile
22: Server certificate verification failed
23: Missing or incorrect password
24: Missing DNS Suffix
25: Missing hostname
26: Missing proxy address and port
27: Missing static IP information
28: Incorrect command line parameters
29: Missing or incorrect network mask
30: Missing or incorrect gateway
31: Missing or incorrect primary DNS
32: Missing or incorrect secondary DNS
33: Invalid Parameter Combination

### (70-99) Connection Errors
70: RPS authentication failed
71: AMT connection failed
72: OS network interfaces lookup failed

### (100-149) Activation, and configuration errors
100: AMT authentication failed
101: WSMAN message error
102: Activation failed
103: Network configuration failed
104: CIRA configuration failed
105: TLS configuration failed
106: WiFi configuration failed
107: AMT features configuration failed
108: 802.1x configuration failed
109: UnableToDeactivate
110: DeactivationFailed
111: UnableToActivate
112: WiFi configuration succeeded with warnings

### (150-199) Maintenance Errors
150: Clock synchronization failed
151: Hostname synchronization failed
152: IP synchronization failed
153: Change password failed

**Comment by graikhel-intel** [comment]
12: Server certificate verification failed ---> It can fall into 2 categories, input error or if a customer is not using Web Cert signed by trusted CA.

**Comment by graikhel-intel** [comment]
**(10-19)** or **(10-29)** --- Input errors to RPC?

**Reply by rsdmike** [reply]
addressed

**Comment by graikhel-intel** [comment]
@rsdmike What do you think about having more reserved numbers for "Input errors to RPC" for future? With Maintenance and other new features, I think we might have more input errors coming up

**Reply by rsdmike** [reply]
i've update the list -- let me know what you think

**Comment by barneytron** [comment]
Is it possible for more than one error to occur during an execution?

**Reply by matt-primrose** [reply]
Should only return 1 error.  Once an error occurs, we stop

**Comment by bgmckown** [comment]
This listing of error codes looks good to me.

**Comment by tim-shockley** [comment]
Edited the list to align with existing code definitions
33: Invalid Parameter Combination
109: UnableToDeactivate
110: DeactivationFailed
111: UnableToActivate
112: WiFi configuration succeeded with warnings
153: Change password failed

note: non-local execution of activation, deactivation, maintenance (codes 100 - 199)  are not currently returned
